x/emissions/types/validations.go

```diff
@@ -1,18 +1,764 @@
+	if err := ValidateTopicId(stakeRemovalInfo.TopicId); err != nil {
+		return errors.Wrap(err, "stakeRemovalInfo topic id is invalid")
+	}
+	if err := ValidateBech32(stakeRemovalInfo.Reputer); err != nil {
+		return errors.Wrap(err, "stakeRemovalInfo reputer is invalid")
+	}
+	if err := ValidateSdkIntRepresentingMonetaryValue(stakeRemovalInfo.Amount); err != nil {
+		return errors.Wrap(err, "stakeRemovalInfo amount is invalid")
+	}
+	if err := ValidateBlockHeight(stakeRemovalInfo.BlockRemovalStarted); err != nil {
+		return errors.Wrap(err, "stakeRemovalInfo block removal started is invalid")
+	}
+	if err := ValidateBlockHeight(stakeRemovalInfo.BlockRemovalCompleted); err != nil {
+		return errors.Wrap(err, "stakeRemovalInfo block removal completed is invalid")
+	}
+	if stakeRemovalInfo.BlockRemovalStarted > stakeRemovalInfo.BlockRemovalCompleted {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest,
+			"stakeRemovalInfo block removal started cannot be greater than block removal completed")
+	}
+	return nil
+}
+
+// validate that a DelegateStakeRemovalInfo follows expected format
+func (delegateStakeRemovalInfo *DelegateStakeRemovalInfo) Validate() error {
+	if err := ValidateTopicId(delegateStakeRemovalInfo.TopicId); err != nil {
+		return errors.Wrap(err, "delegateStakeRemovalInfo topic id is invalid")
+	}
+	if err := ValidateBech32(delegateStakeRemovalInfo.Delegator); err != nil {
+		return errors.Wrap(err, "delegateStakeRemovalInfo delegator is invalid")
+	}
+	if err := ValidateBech32(delegateStakeRemovalInfo.Reputer); err != nil {
+		return errors.Wrap(err, "delegateStakeRemovalInfo reputer is invalid")
+	}
+	if err := ValidateSdkIntRepresentingMonetaryValue(delegateStakeRemovalInfo.Amount); err != nil {
+		return errors.Wrap(err, "delegateStakeRemovalInfo amount is invalid")
+	}
+	if err := ValidateBlockHeight(delegateStakeRemovalInfo.BlockRemovalStarted); err != nil {
+		return errors.Wrap(err, "delegateStakeRemovalInfo block removal started is invalid")
+	}
+	if err := ValidateBlockHeight(delegateStakeRemovalInfo.BlockRemovalCompleted); err != nil {
+		return errors.Wrap(err, "delegateStakeRemovalInfo block removal completed is invalid")
+	}
+	if delegateStakeRemovalInfo.BlockRemovalStarted > delegateStakeRemovalInfo.BlockRemovalCompleted {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest,
+			"delegateStakeRemovalInfo block removal started cannot be greater than block removal completed")
+	}
+	return nil
+}
+
+// validate that a OffchainNode follows the expected format
+func (oc *OffchainNode) Validate() error {
+	if err := ValidateBech32(oc.NodeAddress); err != nil {
+		return errors.Wrap(err, "offchain node node address is invalid")
+	}
+	if err := ValidateBech32(oc.Owner); err != nil {
+		return errors.Wrap(err, "offchain node owner is invalid")
+	}
+	return nil
+}
+
+/// PROTOBUF MESSAGE VALIDATIONS
+
+// validate that a register request follows the expected format
+func (msg *RegisterRequest) Validate() error {
```

Comment:
note for future: wonder how we can write linter to check that every field of every struct was validated. tangentially realted to exhaustruct PR